lumicks/pylake/kymotracker/detail/denoising.py

```diff
@@ -0,0 +1,412 @@
+import numpy as np
+import scipy.signal
+from scipy.ndimage import convolve
+
+
+def equal_length(kernels):
+    """Make the kernels the same length (while keeping them centered)
+
+    Parameters
+    ----------
+    kernels : list of np.ndarray
+        List of kernels to symmetrize.
+
+    Raises
+    ------
+    ValueError
+        if any of the kernel isn't odd sized (violates the ability to center it).
+    """
+    target_length = len(kernels[-1])
+
+    def pad(kernel):
+        if len(kernel) % 2 == 0:
+            raise ValueError("This function should only be used with odd sized kernels")
+
+        padding = np.zeros(((target_length - len(kernel)) // 2, 1))
+        return np.vstack((padding, kernel, padding))
+
+    return [pad(k) for k in kernels]
+
+
+def generate_product_filters(kernels):
+    """Generate product filters
+
+    Generates a list of kernels that consist of the previous kernels convolved.
+
+    Parameters
+    ----------
+    kernels : list of np.ndarray
+        List of 1-dimensional kernels to calculate the convolutional product of.
+    """
+    current = np.asarray([[1.0]])
+    convolved_kernels = [current]
+    for k in kernels:
+        current = scipy.signal.convolve(current, k, mode="full")
+        convolved_kernels.append(current)
+
+    return convolved_kernels
+
+
+def generate_bspline_kernels(num_kernels):
+    """Returns a list of à trous b3 spline kernels.
+
+    Parameters
+    ----------
+    num_kernels : int
+        Number of kernels we want
+
+    Returns
+    -------
+    kernels : list of np.ndarray
+        1-dimensional kernels of varying sizes
+    """
+    k = np.arange(1, num_kernels + 1)
+    base_kernel = np.array([1 / 16, 1 / 4, 3 / 8, 1 / 4, 1 / 16])
+    kernels = []
+    for zeros in 2 ** (k - 1) - 1:
+        kernel = np.zeros(len(base_kernel) + zeros * (len(base_kernel) - 1))
+        kernel[:: zeros + 1] = base_kernel
+        kernels.append(kernel.reshape((-1, 1)))
+
+    return kernels
+
+
+def tau(kernel, power):
+    """Calculate filter summation from Sec 2. Paragraph 1."""
+    return np.sum(kernel**power)
+
+
+def calculate_vst_coefficients(kernel):
+    """Calculate optimal transform coefficients and expected standard deviation.
+
+    We calculate b and c for the following transform:
+
+        Z = b sign(Y + c) * sqrt(Y + c)
+
+    Using the coefficients from Eq. 10 and b above Eq. 12 we estimate a transform such that
```

Comment:
it would be useful here to state which paper the specific eq numbers are from. I don't think you need to repeat the full citation again, maybe just `(Zhang 2008)` would be good enough? still would be really nice if we could have a global reference list for the module 😞 

on closer looks, also seems like the eq numbers might be wrong? I think you're using slightly different terminology for the variables or am I missing something?